### PR TITLE
Update controller-runtime and dependencies

### PR DESCRIPTION
**Pull request by bshephar**
This change updates the controller-runtime and associated dependencies. We fix the webhook function signatures, the Watches() implementations and the webhook declaration when the manager is initialized to comply with the latest CR standards.

Depends-On: https://github.com/openstack-k8s-operators/ci-framework/pull/954
Depends-On: https://github.com/openstack-k8s-operators/install_yamls/pull/730

### CODE REVIEW

**Comment by bshephar** [comment]
This will fail go vet because of the million dependencies. But it might just be a case of overriding each of them until we can get all of this merged.

**Comment by bshephar** [comment]
Github check failing because it's still running go 1.19:
```
go: go.mod file indicates go 1.20, but maximum version supported by tidy is 1.19
```

**Comment by bshephar** [comment]
recheck

**Comment by bshephar** [comment]
Running the make bundle locally to check the kuttl error. But skopeo works:
```
❯ skopeo copy docker://quay.io/openstack-k8s-operators/placement-operator-bundle:12812f6d9604fd781a593590c8250eb8a481fec6 dir:tmp/bundle_extract/tmp
Getting image source signatures
Copying blob 472bb3aeebd9 done   |
Copying config 32e06924d1 done   |
Writing manifest to image destination
```

I guess the Error in this string is actually the problem, but I don't immediately see where that come from:
```
+ for BUNDLE in $(hack/pin-bundle-images.sh | tr "," " ")
+ skopeo copy docker://quay.io/openstack-k8s-operators/placement-operator-bundle:12812f6d9604fd781a593590c8250eb8a481fec6Error: dir:tmp/bundle_extract/tmp
time="2024-02-15T02:03:09Z" level=fatal msg="Invalid source name docker://quay.io/openstack-k8s-operators/placement-operator-bundle:12812f6d9604fd781a593590c8250eb8a481fec6Error:: invalid reference format"

```

**Comment by bshephar** [comment]
Reproduced:
```
❯ for BUNDLE in $(hack/pin-bundle-images.sh | tr "," " "); do echo $BUNDLE; done
quay.io/openstack-k8s-operators/barbican-operator-bundle:c2163adf7b2dbe486f6ae14b0e17a8ba85a8dac9
quay.io/openstack-k8s-operators/cinder-operator-bundle:8d723e605478f874ba6fa764b7011c01662f1a0d
quay.io/openstack-k8s-operators/dataplane-operator-bundle:56156b3ceaa2b719ea6deafb14b3060f4b8dd073
quay.io/openstack-k8s-operators/designate-operator-bundle:ae268f19ceffa2636f87a1289181cab75eef1419
quay.io/openstack-k8s-operators/glance-operator-bundle:457fe24ee6ff347f6479bed2f5f1c72306105824
quay.io/openstack-k8s-operators/heat-operator-bundle:b1c60ab46d1f3b40e3142d99034113faaebf1c04
quay.io/openstack-k8s-operators/horizon-operator-bundle:e28d5c7d95a88a57a91ada568aab9ea2af59a006
quay.io/openstack-k8s-operators/infra-operator-bundle:179defb96a3363bc3684f080ea30f8e59de68017
quay.io/openstack-k8s-operators/ironic-operator-bundle:c87a0c959dc803827927f99a414ed0ae82fc4f71
quay.io/openstack-k8s-operators/keystone-operator-bundle:55af8e58473d6311c7a645b8ae87d8c9bc5b01f8
quay.io/openstack-k8s-operators/manila-operator-bundle:c4857b427c2d590ee49cdf419eb514e22f237cb1
quay.io/openstack-k8s-operators/mariadb-operator-bundle:3b28b9d3d09ba3587644f828281bb2241ecedef2
quay.io/openstack-k8s-operators/neutron-operator-bundle:985ddd95c48be5268b815ca9601a3c799b927c86
quay.io/openstack-k8s-operators/nova-operator-bundle:a1f8067ef5c629e24886de382cb1d3da53da16ba
quay.io/openstack-k8s-operators/octavia-operator-bundle:c57a33f4e98ad612c3b8a685a6928e00ee9821c3
quay.io/openstack-k8s-operators/openstack-ansibleee-operator-bundle:9bf047b38c40affa213b4d7ef1ef5b99799ef095
quay.io/openstack-k8s-operators/openstack-baremetal-operator-bundle:80b4f87cfc2b9519f2eb7b05f48ae897b1eeb56f
quay.io/openstack-k8s-operators/ovn-operator-bundle:93933dc002b7f19d094b3004aab48b29095509a0
quay.io/openstack-k8s-operators/placement-operator-bundle:12812f6d9604fd781a593590c8250eb8a481fec6Error:
SHA
is
empty
Could
not
find
container
with
tag
9615ae5dbf3c
in
https://quay.io/api/v1/repository/openstack-k8s-operators
```

Added some debugging here, seems to be the Swift image with problems:
```
hack/pin-bundle-images.sh

,quay.io/openstack-k8s-operators/barbican-operator-bundle:c2163adf7b2dbe486f6ae14b0e17a8ba85a8dac9,quay.io/openstack-k8s-operators/cinder-operator-bundle:8d723e605478f874ba6fa764b7011c01662f1a0d,quay.io/openstack-k8s-operators/dataplane-operator-bundle:56156b3ceaa2b719ea6deafb14b3060f4b8dd073,quay.io/openstack-k8s-operators/designate-operator-bundle:ae268f19ceffa2636f87a1289181cab75eef1419,quay.io/openstack-k8s-operators/glance-operator-bundle:457fe24ee6ff347f6479bed2f5f1c72306105824,quay.io/openstack-k8s-operators/heat-operator-bundle:b1c60ab46d1f3b40e3142d99034113faaebf1c04,quay.io/openstack-k8s-operators/horizon-operator-bundle:e28d5c7d95a88a57a91ada568aab9ea2af59a006,quay.io/openstack-k8s-operators/infra-operator-bundle:179defb96a3363bc3684f080ea30f8e59de68017,quay.io/openstack-k8s-operators/ironic-operator-bundle:c87a0c959dc803827927f99a414ed0ae82fc4f71,quay.io/openstack-k8s-operators/keystone-operator-bundle:55af8e58473d6311c7a645b8ae87d8c9bc5b01f8,quay.io/openstack-k8s-operators/manila-operator-bundle:c4857b427c2d590ee49cdf419eb514e22f237cb1,quay.io/openstack-k8s-operators/mariadb-operator-bundle:3b28b9d3d09ba3587644f828281bb2241ecedef2,quay.io/openstack-k8s-operators/neutron-operator-bundle:985ddd95c48be5268b815ca9601a3c799b927c86,quay.io/openstack-k8s-operators/nova-operator-bundle:a1f8067ef5c629e24886de382cb1d3da53da16ba,quay.io/openstack-k8s-operators/octavia-operator-bundle:c57a33f4e98ad612c3b8a685a6928e00ee9821c3,quay.io/openstack-k8s-operators/openstack-ansibleee-operator-bundle:9bf047b38c40affa213b4d7ef1ef5b99799ef095,quay.io/openstack-k8s-operators/openstack-baremetal-operator-bundle:80b4f87cfc2b9519f2eb7b05f48ae897b1eeb56f,quay.io/openstack-k8s-operators/ovn-operator-bundle:93933dc002b7f19d094b3004aab48b29095509a0,quay.io/openstack-k8s-operators/placement-operator-bundle:12812f6d9604fd781a593590c8250eb8a481fec6Base = swift ImageBase = \n
Error: SHA is empty, Could not find container with tag 9615ae5dbf3c in https://quay.io/api/v1/repository/openstack-k8s-operators
```

Looks like it failed to merge with some go.sum issues:
https://github.com/openstack-k8s-operators/swift-operator/actions/runs/7908400561

pre-commit doesn't raise any concerns now. Submitting a test PR to swift to test things.

**Comment by bshephar** [comment]
recheck

**Comment by bshephar** [comment]
recheck

**Comment by raukadah** [comment]
recheck after node hold

**Comment by raukadah** [comment]
recheck

**Comment by raukadah** [comment]
@bshephar current failure at [make bundle_build](https://logserver.rdoproject.org/51/651/906d62d1eb6c5e918eca95c28c0cc7d6dbbcd135/github-check/openstack-k8s-operators-content-provider/f3faa2f/ci-framework-data/logs/ci_script_005_openstack-operator_-_call_bundle-build.log)
```
podman build -f custom-bundle.Dockerfile -t 38.102.83.65:5001/openstack-k8s-operators/openstack-operator-bundle:906d62d1eb6c5e918eca95c28c0cc7d6dbbcd135 .
[1/3] STEP 1/9: FROM golang:1.19 AS builder
Error: creating build container: short-name resolution enforced but cannot prompt without a TTY
make: *** [Makefile:271: bundle-build] Error 125
```
we need to update GOLANG_CTX to 1.20 here https://github.com/openstack-k8s-operators/openstack-operator/blob/main/custom-bundle.Dockerfile#L1

**Comment by rabi** [comment]
Looks like telemetry bundle is still missing.. It's there in quay now...

**Comment by bshephar** [comment]
I still had a `replace` for telemetry-operator still. I've dropped that now.

File: apis/go.mod [hunk 1]
```diff
@@ -1,112 +1,104 @@
...
+	github.com/openstack-k8s-operators/nova-operator/api v0.3.1-0.20240214174908-a1f8067ef5c6
+	github.com/openstack-k8s-operators/octavia-operator/api v0.3.1-0.20240215003649-c57a33f4e98a
+	github.com/openstack-k8s-operators/ovn-operator/api v0.3.1-0.20240214165507-93933dc002b7
+	github.com/openstack-k8s-operators/placement-operator/api v0.3.1-0.20240214182711-12812f6d9604
+	github.com/openstack-k8s-operators/swift-operator/api v0.3.1-0.20240215033724-efcba239ce14
+	github.com/openstack-k8s-operators/telemetry-operator/api v0.3.1-0.20240215051331-c5c5ad24987b
 	github.com/rabbitmq/cluster-operator/v2 v2.5.0
-	k8s.io/apimachinery v0.27.7
-	sigs.k8s.io/controller-runtime v0.15.1
+	k8s.io/apimachinery v0.29.0
```
Comment (gibizer) [inline apis/go.mod:28]: that sounds wrong

File: go.mod [hunk 2]
```diff
@@ -9,52 +9,53 @@ require (
...
 	github.com/rabbitmq/cluster-operator/v2 v2.5.0
 	go.uber.org/zap v1.26.0
-	golang.org/x/exp v0.0.0-20240119083558-1b970713d09a
-	k8s.io/api v0.27.7
-	k8s.io/apimachinery v0.27.7
-	k8s.io/client-go v0.27.7
-	sigs.k8s.io/controller-runtime v0.15.1
+	golang.org/x/exp v0.0.0-20240213143201-ec583247a57a
+	k8s.io/api v0.28.6
+	k8s.io/apimachinery v0.29.0
```
Comment (gibizer) [inline go.mod:42]: that should be 0.28

**Review by gibizer — CHANGES_REQUESTED** [review]

**Review by gibizer — APPROVED** [review]
If CI is green the we can do the follow up to pull back `k8s.io/apimachinery v0.29.0` separately

File: apis/go.mod (hunk 1 above)
Comment (stuggi) [reply apis/go.mod]: changed in latest commit

File: go.mod (hunk 2 above)
Comment (stuggi) [reply go.mod]: changed in latest commit

**Comment by gibizer** [comment]
recheck ci-framework PR needed a rebase

**Review by stuggi — APPROVED** [review]
/lgtm

**Pull request merged**